### PR TITLE
Multi amplitude embedding with transform

### PR DESCRIPTION
**Pull request by KetpuntoG**
**Context:** In certain situations it may be convenient to be able to use more than one `AmplitudeEmbedding` in the same circuit. A first approximation has been made but it has been decided to use circuit transformations to merge these gates. [(#1890)](https://github.com/PennyLaneAI/pennylane/pull/1890)

**Description of the Change:** A `merge_amplitude_embedding` transformation has been created, which takes a quantum function and returns it transformed by joining all gates of this type together.

**Example:**

```python
import pennylane as qml
from pennylane.transforms import merge_amplitude_embedding

@merge_amplitude_embedding
def qfunc():
    
    qml.AmplitudeEmbedding([0,1,0,0], wires = [0,1])
    qml.AmplitudeEmbedding([0,1], wires = 2)
    
    return qml.expval(qml.PauliZ(wires = 0))

dev = qml.device("default.qubit", wires = 3)
qnode = qml.QNode(qfunc, dev)
print(qml.draw(qnode)())
```

 0: ──╭AmplitudeEmbedding(M0)──┤ ⟨Z⟩ 
 1: ──├AmplitudeEmbedding(M0)──┤     
 2: ──╰AmplitudeEmbedding(M0)──┤     
M0 =
[0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]

**Benefits:**

Convenience when programming some routines such as the calculation of the inner product through the swap test.

**Possible Drawbacks:**

**Related GitHub Issues:**
[(#1140)](https://github.com/PennyLaneAI/pennylane/issues/1140)

### CODE REVIEW

**Comment by KetpuntoG** [comment]
[sc-11240]

File: tests/transforms/test_optimization/test_merge_amplitude_embedding.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+        dev = qml.device("default.qubit", wires=2)
+        qnode = qml.QNode(transformed_qfunc, dev)
+
+        with pytest.raises(DeviceError, match="applied in the same qubit"):
+            qnode()
+
+
+class TestMergeAmplitudeEmbeddingInterfaces:
+    """Test that amplitude merging works in all interfaces."""
```
Comment (antalszava) [inline tests/transforms/test_optimization/test_merge_amplitude_embedding.py:63]:
```suggestion
    """Test that merging amplitude embedding operations works in all interfaces."""
```

File: tests/transforms/test_optimization/test_merge_amplitude_embedding.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            qml.Hadamard(wires=0)
+            qml.state()
+
+        transformed_qfunc = merge_amplitude_embedding(qfunc)
+        ops = qml.transforms.make_tape(transformed_qfunc)().operations
+
+        assert len(ops) == 3
+
+        # Check that the solution is as expected.
+
```
Comment (antalszava) [inline tests/transforms/test_optimization/test_merge_amplitude_embedding.py:42]:
```suggestion
```

File: tests/transforms/test_optimization/test_merge_amplitude_embedding.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import pennylane as qml
+from pennylane.transforms.optimization import merge_amplitude_embedding
+from pennylane._device import DeviceError
+
+
+class TestMergeAmplitudeEmbedding:
+    """Test that amplitude embedding gates are combined into a single."""
+
+    def test_multi_amplitude_embedding(self):
+        """Test that the transformation is working correctly by joining two amplitudeEmbedding."""
```
Comment (antalszava) [inline tests/transforms/test_optimization/test_merge_amplitude_embedding.py:27]:
```suggestion
        """Test that the transformation is working correctly by joining two AmplitudeEmbedding."""
```

**Review by antalszava — CHANGES_REQUESTED** [review]
Hi @KetpuntoG, this is amazing, looks great! Congrats and thank you for hanging in there through the extra required changes :blush: :tada: 

Nothing major for the PR, one test case that I think could be useful to add is a case where a QNode is being decorated and it's checked that evaluating it works as expected (and as proposed in the original issue).

File: tests/transforms/test_optimization/test_merge_amplitude_embedding.py [hunk 4]
```diff
@@ -58,9 +57,23 @@ def qfunc():
...
+        @merge_amplitude_embedding
+        def qfunc():
+            qml.AmplitudeEmbedding([0, 1, 0, 0], wires=[0, 1])
+            qml.AmplitudeEmbedding([0, 1], wires=2)
+
+            return qml.state()
+
+        dev = qml.device("default.qubit", wires=3)
+        qnode = qml.QNode(qfunc, dev)
+        assert qnode()[3] == 1.0
```
Comment (antalszava) [inline tests/transforms/test_optimization/test_merge_amplitude_embedding.py:72]:
```suggestion
        dev = qml.device("default.qubit", wires=3)
        
        @merge_amplitude_embedding
        @qml.qnode(dev)
        def qnode():
            qml.AmplitudeEmbedding([0, 1, 0, 0], wires=[0, 1])
            qml.AmplitudeEmbedding([0, 1], wires=2)

            return qml.state()

        assert qnode()[3] == 1.0
```

**Review by antalszava — APPROVED** [review]
Looks great! :partying_face: Thank you again for the great solution. :tada:

**Pull request merged**